File: week-4/all-possible-full-binary-tree/all-possible-full-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number} N
 * @return {TreeNode[]}
 */
const memo = {};
var allPossibleFBT = function(N) {
    if (N===0) return [];
    if (N===1) return [new TreeNode(0)];
    
    if (N in memo) return memo[N];
    
    const res = [];
    for (let i=0; i<N; i++) {
        const j = N-1-i;
        for (let left of allPossibleFBT(i)) {
            for (let right of allPossibleFBT(j)) {
                const temp = new TreeNode(0);
                temp.left = left;
                temp.right = right;
                res.push(temp);
            }
        }
    }
    
    memo[N] = res;
    return memo[N];
};

File: week-4/all-possible-full-binary-tree/all-possible-full-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def __init__(self):
        self.memo = {}
        
    def allPossibleFBT(self, N: int) -> List[TreeNode]:
        if N == 0: return []
        if N == 1: return [TreeNode(0)]
        
        if N in self.memo: return self.memo[N]
        
        res = []
        for i in range(N):
            j = N-1-i
            for left in self.allPossibleFBT(i):
                for right in self.allPossibleFBT(j):
                    temp = TreeNode(0)
                    temp.left = left
                    temp.right = right
                    res.append(temp)
        
        self.memo[N] = res
        return self.memo[N]
